service/worker/pernamespaceworker.go

```diff
@@ -260,24 +275,43 @@ func (wm *perNamespaceWorkerManager) removeWorker(ns *namespace.Namespace) {
 	delete(wm.workers, ns.ID())
 }
 
-func (wm *perNamespaceWorkerManager) getWorkerMultiplicity(ns *namespace.Namespace) (int, int, error) {
+func (wm *perNamespaceWorkerManager) getWorkerAllocation(ns *namespace.Namespace) (*workerAllocation, error) {
+	desiredWorkersCount, err := wm.getConfiguredWorkersCountFor(ns)
+	if err != nil {
+		return nil, err
+	}
+	if desiredWorkersCount == 0 {
+		return &workerAllocation{0, 0}, nil
+	}
+	localCount, err := wm.getLocallyDesiredWorkersCount(ns, desiredWorkersCount)
+	if err != nil {
+		return nil, err
+	}
+	return &workerAllocation{desiredWorkersCount, localCount}, nil
+}
+
+func (wm *perNamespaceWorkerManager) getConfiguredWorkersCountFor(ns *namespace.Namespace) (uint16, error) {
 	totalWorkers := wm.config.PerNamespaceWorkerCount(ns.Name().String())
-	// This can result in fewer than the intended number of workers if totalWorkers > 1, because
-	// multiple lookups might land on the same node. To compensate, we increase the number of
-	// pollers in that case, but it would be better to try to spread them across our nodes.
-	// TODO: implement this properly using LookupN in serviceResolver
-	multiplicity := 0
-	for i := 0; i < totalWorkers; i++ {
-		key := fmt.Sprintf("%s/%d", ns.ID().String(), i)
-		target, err := wm.serviceResolver.Lookup(key)
-		if err != nil {
-			return 0, 0, err
-		}
-		if target.Identity() == wm.self.Identity() {
-			multiplicity++
-		}
+	if totalWorkers < 0 || totalWorkers > math.MaxUint16 {
+		err := fmt.Errorf("%w namespace %s, workers count %d", dynamicconfig.ErrInvalidConfiguration, ns.Name(), totalWorkers)
+		return 0, err
 	}
-	return multiplicity, totalWorkers, nil
+	return uint16(totalWorkers), nil
+}
+
+func (wm *perNamespaceWorkerManager) getLocallyDesiredWorkersCount(ns *namespace.Namespace, desiredNumberOfWorkers uint16) (uint16, error) {
+	key := ns.ID().String()
+	hostsInNamespace := wm.serviceResolver.LookupN(key, desiredNumberOfWorkers)
+	hostsCount := uint16(len(hostsInNamespace))
+	if hostsCount == 0 {
+		return 0, membership.ErrInsufficientHosts
+	}
+	maxWorkersPerHost := int(math.Ceil(float64(desiredNumberOfWorkers) / float64(hostsCount)))
```

Comment:
integer math seems simpler? (and you can overshoot)
```suggestion
	maxWorkersPerHost := desiredNumberOfWorkers/hostsCount + 1
```

Reply:
👍 Thank you!